### PR TITLE
Nr 318119 auto collect stdout & stderr logs

### PR DESCRIPTION
**Pull request by mbruin-NR**
https://new-relic.atlassian.net/browse/NR-318119

### CODE REVIEW

File: Agent/Utilities/NRLogger.m [hunk 1]
```diff
@@ -346,7 +357,6 @@ - (NSData*) jsonDictionary:(NSDictionary*)message {
 
 
     NSMutableDictionary *providedAttributes = [message mutableCopy];
-    [providedAttributes removeObjectsForKeys:@[NRLogMessageLevelKey,NRLogMessageFileKey,NRLogMessageLineNumberKey,NRLogMessageMethodKey,NRLogMessageTimestampKey,NRLogMessageMessageKey]];
```
Comment (cdillard-NewRelic) [inline Agent/Utilities/NRLogger.m:349]: The intention of this line is to remove reserved attributes from user provided attributes when passed to NewRelic.logAttributes.

Comment (cdillard-NewRelic) [reply Agent/Utilities/NRLogger.m:349]: Do you still want to remove this line? If so can you explain why?

File: Test Harness/NRTestApp/NRTestApp/AppDelegate.swift [hunk 2]
```diff
@@ -33,7 +33,9 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
                                  NRMAFeatureFlags.NRFeatureFlag_OfflineStorage])
         // Note: Disabled by default. Enable or disable (default) flag to enable background reporting.
         // NewRelic.enableFeatures([NRMAFeatureFlags.NRFeatureFlag_BackgroundReporting])
-
+        
+        // Note: Disabled by default. Enable or disable (default) flag to enable auto collect of stdout.
+        //NewRelic.enableFeatures([NRMAFeatureFlags.NRFeatureFlag_RedirectStdOutStdErr])
```
Comment (cdillard-NewRelic) [inline Test Harness/NRTestApp/NRTestApp/AppDelegate.swift:38]: Sounds like John  would like AutoCollectedLogs to be on by default. But we should double check with him and Android.

File: Agent/Utilities/NRLogger.m (hunk 1 above)
Comment (mbruin-NR) [reply Agent/Utilities/NRLogger.m]: This line doesn't interfere with auto collected logs like I originally thought. Adding it back it.

File: Agent/FeatureFlags/NRMAFlags.m [hunk 3]
```diff
@@ -47,7 +47,8 @@ + (NRMAFeatureFlags) featureFlags
                               NRFeatureFlag_NetworkRequestEvents | 
                               NRFeatureFlag_RequestErrorEvents |
                               NRFeatureFlag_DistributedTracing |
-                              NRFeatureFlag_AppStartMetrics;
+                              NRFeatureFlag_AppStartMetrics |
+                              NRFeatureFlag_AutoCollectLogs;
```
Comment (cthomas-newrelic) [inline Agent/FeatureFlags/NRMAFlags.m:51]: Careful with new feature flags: we _try_ to keep FF parity between the agents.

File: Agent/Utilities/NRAutoLogCollector.h [hunk 4]
```diff
@@ -0,0 +1,19 @@
+//
+//  NRAutoLogCollector.h
+//  Agent
+//
+//  Created by Mike Bruin on 10/9/24.
```
Comment (cthomas-newrelic) [inline Agent/Utilities/NRAutoLogCollector.h:5]: Generally, we try to avoid putting our names into commits. It's on the commit already. If it's a team practice, I understand.

File: Agent/Utilities/NRAutoLogCollector.m [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+@implementation NRAutoLogCollector
+
++ (BOOL) redirectStandardOutputAndError {
+    if (hasRedirectedStdOut){
+        return true;
+    }
+    // Create pipes for stdout and stderr
+    if (pipe(stdoutPipe) == -1 || pipe(stderrPipe) == -1) {
+        return false;
```
Comment (cthomas-newrelic) [inline Agent/Utilities/NRAutoLogCollector.m:30]: Should close the valid pipe if the other returns -1.

File: Agent/Utilities/NRAutoLogCollector.m [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        close(stdoutPipe[1]);
+        close(stderrPipe[0]);
+        close(stderrPipe[1]);
+        close(saved_stdout);
+        close(saved_stderr);
+        return false;
+    }
+    close(stdoutPipe[1]); // Close the original write end of the stdout pipe
+    close(stderrPipe[1]); // Close the original write end of the stderr pipe
+
```
Comment (cthomas-newrelic) [inline Agent/Utilities/NRAutoLogCollector.m:56]: You should investigate disabling buffering of the pipes

**Review by cthomas-newrelic — COMMENTED** [review]
Nice job.

File: Agent/Utilities/NRAutoLogCollector.h (hunk 4 above)
Comment (mbruin-NR) [reply Agent/Utilities/NRAutoLogCollector.h:5]: It is team practice, mainly because Xcode automatically adds it when you create a new file.

File: Agent/FeatureFlags/NRMAFlags.m (hunk 3 above)
Comment (mbruin-NR) [reply Agent/FeatureFlags/NRMAFlags.m:51]: I think it's very important to have this feature flag for iOS even if it's not on Android. This is because it breaks logging to Xcode console. If we want to keep parity I would greatly appriciate Android adding the feature flag and ability to disable auto capturing logs.

File: Agent/Utilities/NRAutoLogCollector.m (hunk 6 above)
Comment (cthomas-newrelic) [reply Agent/Utilities/NRAutoLogCollector.m:61]: Yes, I believe that's correct.

Comment (mbruin-NR) [reply Agent/Utilities/NRAutoLogCollector.m:61]: I tried adding that but it seems to mess with the accuracy of the logs. Giving more logs out of order and having partial logs.

File: Agent/FeatureFlags/NRMAFlags.m (hunk 3 above)
Comment (smalsam-newr) [reply Agent/FeatureFlags/NRMAFlags.m:51]: Instead of a feature flag, could we detect when the debugger is attached? I imagine we wouldn't want to collect logs in that instance

Comment (cdillard-NewRelic) [reply Agent/FeatureFlags/NRMAFlags.m]: I think we still need the feature flag, and I believe this auto capture should be disabled by default.  It should also not collect logs when the debugger is attached.

Comment (cdillard-NewRelic) [reply Agent/FeatureFlags/NRMAFlags.m]: The reason being that we like to gate all new features with a feature flag in case there is issues.

File: Agent/Utilities/NRLogger.m [hunk 7]
```diff
@@ -299,19 +316,37 @@ - (NSData*) jsonDictionary:(NSDictionary*)message {
...
-                                         NRLogMessageEntityGuidKey: entityGuid,                                                                                                  // 9
-                                         NRLogMessageInstrumentationProviderKey: NRLogMessageMobileValue,                                                                        // 10
-                                         NRLogMessageInstrumentationNameKey: nativePlatform,                                                                                               // 11
-                                         NRLogMessageInstrumentationVersionKey: [NRMAAgentConfiguration connectionInformation].deviceInformation.agentVersion,                   // 12
-                                         NRLogMessageInstrumentationCollectorKey: nativePlatform};                                                                                         // 13
+    NSMutableDictionary *requiredAttributes = [NSMutableDictionary dictionary];
```
Comment (cdillard-NewRelic) [inline Agent/Utilities/NRLogger.m:319]: Why was this changed? I am working on some changes in this area. Im concerned that we might miss required attributes with the way this was changed.

File: Agent/Utilities/NRLogger.m [hunk 8]
```diff
@@ -299,19 +316,37 @@ - (NSData*) jsonDictionary:(NSDictionary*)message {
...
+
+    value = [[message objectForKey:NRLogMessageMessageKey]stringByReplacingOccurrencesOfString:@"\"" withString:@"\\\""];
+    if (value) [requiredAttributes setObject:value forKey:NRLogMessageMessageKey]; // 6
+
+    if (NRSessionId) [requiredAttributes setObject:NRSessionId forKey:NRLogMessageSessionIdKey]; // 7
+    if (nrAppId) [requiredAttributes setObject:nrAppId forKey:NRLogMessageAppIdKey]; // 8
+    if (entityGuid) [requiredAttributes setObject:entityGuid forKey:NRLogMessageEntityGuidKey]; // 9
+
+    [requiredAttributes setObject:NRLogMessageMobileValue forKey:NRLogMessageInstrumentationProviderKey]; // 10
+    [requiredAttributes setObject:nativePlatform forKey:NRLogMessageInstrumentationNameKey]; // 11
```
Comment (cdillard-NewRelic) [inline Agent/Utilities/NRLogger.m:344]:
```suggestion
    [requiredAttributes setObject:name forKey:NRLogMessageInstrumentationNameKey]; // 11
```

File: Agent/Utilities/NRLogger.m (hunk 7 above)
Comment (cdillard-NewRelic) [reply Agent/Utilities/NRLogger.m:319]: If you'd like to keep the changes, which is could be okay if they aren't required by log API... then lets make the following change:

```
                                         // "instrumentation.name"
                                         NRLogMessageInstrumentationNameKey: name, // 11                                                         
                                         // "collector.name
                                         NRLogMessageInstrumentationCollectorKey: nativePlatform};  // 13   
```

Comment (mbruin-NR) [reply Agent/Utilities/NRLogger.m:319]: I made this change so that the agent wouldn't crash if the value was null. There is still probably a change we could make so that it again won't add a log if it doesn't have the required attributes.

**Comment by cdillard-NewRelic** [comment]
Do we think there is anyway to allow the console and auto collected logs to work at the same time? If not its fine but we should note it in the docs. Thanks!

**Comment by mbruin-NR** [comment]
> Do we think there is anyway to allow the console and auto collected logs to work at the same time? If not its fine but we should note it in the docs. Thanks!

I can't think of a way to put it on the console too, maybe we could try to redirect back to console on every log message but that's not how it's setup. It's going to be part of the docs.

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -16,15 +16,15 @@ jobs:
     name : TestIOS 
     # runs-on: will be set to macos-latest when running on actual GHA. 
     # *** runs-on: ubuntu-latest is used when running via act on mac os. ***
-    runs-on: macos-13
```
Comment (smalsam-newr) [inline .github/workflows/main.yml:19]: Are these updates on the runner because of this feature?

**Review by smalsam-newr — APPROVED** [review]

File: .github/workflows/main.yml (hunk 9 above)
Comment (mbruin-NR) [reply .github/workflows/main.yml:19]: Github started failing watchOS and tvOS with the Xcode version it was set to not being present. So I updated it but I know there have been other pr's where that wasn't an issue. Maybe it was fixed on the GitHub side, I can remove it if we want to verify.

**Review by cdillard-NewRelic — APPROVED** [review]

File: Agent/FeatureFlags/NRMAFlags.m (hunk 3 above)
Comment (cdillard-NewRelic) [inline Agent/FeatureFlags/NRMAFlags.m:51]: `NRFeatureFlag_AutoCollectLogs` must be disabled by default. As my resolved comment indicated. Thanks!
